From 95f82155884240bf4608f5c1f2681eb0ce69a38b Mon Sep 17 00:00:00 2001
From: Alex Williamson <alex.williamson@redhat.com>
Date: Fri, 6 Jun 2014 16:27:55 +0200
Subject: [PATCH 1/2] Avoid setting PSH flag when SYN flag is set

RH-Author: Alex Williamson <alex.williamson@redhat.com>
Message-id: <20140606162755.23875.32743.stgit@bling.home>
Patchwork-id: 59170
O-Subject: [RHEL6.6 gpxe PATCH 1/2] [tcp] Avoid setting PSH flag when SYN flag is set
Bugzilla: 1105189
RH-Acked-by: Jeffrey Cody <jcody@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

From: Michael Brown <mcb30@etherboot.org>

Bugzilla: 1105189
Upstream commit: 3ed468e0c58bb0eb94925c18a0461cd5014b5777 (ipxe.git)

Some firewall devices seem to regard SYN,PSH as an invalid flag
combination and reject the packet.  Fix by setting PSH only if SYN is
not set.

Reported-by: DSE Incorporated <dseinc@gmail.com>
---
 src/net/tcp.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

Signed-off-by: jen <jen@redhat.com>
---
 src/net/tcp.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/src/net/tcp.c b/src/net/tcp.c
index 6bcd193..0ef6577 100644
--- a/src/net/tcp.c
+++ b/src/net/tcp.c
@@ -471,6 +471,8 @@ static int tcp_xmit ( struct tcp_connection *tcp, int force_send ) {
 		tsopt->tsopt.tsval = ntohl ( currticks() );
 		tsopt->tsopt.tsecr = ntohl ( tcp->ts_recent );
 	}
+	if ( ! ( flags & TCP_SYN ) )
+		flags |= TCP_PSH;
 	tcphdr = iob_push ( iobuf, sizeof ( *tcphdr ) );
 	memset ( tcphdr, 0, sizeof ( *tcphdr ) );
 	tcphdr->src = tcp->local_port;
@@ -478,7 +480,7 @@ static int tcp_xmit ( struct tcp_connection *tcp, int force_send ) {
 	tcphdr->seq = htonl ( tcp->snd_seq );
 	tcphdr->ack = htonl ( tcp->rcv_ack );
 	tcphdr->hlen = ( ( payload - iobuf->data ) << 2 );
-	tcphdr->flags = ( flags | TCP_PSH );
+	tcphdr->flags = flags;
 	tcphdr->win = htons ( tcp->rcv_win );
 	tcphdr->csum = tcpip_chksum ( iobuf->data, iob_len ( iobuf ) );
 
-- 
1.7.1

